src/types.rs

```diff
@@ -177,3 +177,130 @@ impl<'a> From<&'a QUIC_CONNECTION_EVENT> for ConnectionEvent<'a> {
         }
     }
 }
+
+/// Stream callback events
+pub enum StreamEvent<'a> {
+    StartComplete {
+        status: crate::Status,
+        id: crate::u62,
+        peer_accepted: bool,
+    },
+    Receive {
+        absolute_offset: u64,
+        total_buffer_length: &'a mut u64,       // out parameter
+        buffers: &'a [crate::ffi::QUIC_BUFFER], // TODO: impl buffer wrapper types
+        flags: crate::ffi::QUIC_RECEIVE_FLAGS,
+    },
+    SendComplete {
+        cancelled: bool,
+        client_context: *const std::ffi::c_void,
+    },
+    PeerSendShutdown,
+    PeerSendAborted {
+        error_code: crate::u62,
+    },
+    PeerReceiveAborted {
+        error_code: crate::u62,
+    },
+    SendShutdownComplete {
+        graceful: bool,
+    },
+    ShutdownComplete {
+        connection_shutdown: bool,
+        app_close_in_progress: bool,
+        connection_shutdown_by_app: bool,
+        connection_closed_remotely: bool,
+        connection_error_code: crate::u62,
+        connection_close_status: crate::Status,
+    },
+    IdealSendBufferSize {
+        byte_count: u64,
+    },
+    PeerAccepted,
+    CancelOnLoss {
+        error_code: &'a mut crate::u62, // out param
+    },
+}
+
+impl<'b> From<&'b mut crate::ffi::QUIC_STREAM_EVENT> for StreamEvent<'b> {
+    fn from(value: &'b mut crate::ffi::QUIC_STREAM_EVENT) -> Self {
+        match value.Type {
+            crate::ffi::QUIC_STREAM_EVENT_TYPE_QUIC_STREAM_EVENT_START_COMPLETE => {
+                let ev = unsafe { value.__bindgen_anon_1.START_COMPLETE };
+                Self::StartComplete {
+                    status: crate::Status(ev.Status),
+                    id: ev.ID,
+                    peer_accepted: ev.PeerAccepted() != 0,
+                }
+            }
+            crate::ffi::QUIC_STREAM_EVENT_TYPE_QUIC_STREAM_EVENT_RECEIVE => {
+                let ev = unsafe { &mut value.__bindgen_anon_1.RECEIVE };
+                Self::Receive {
+                    absolute_offset: ev.AbsoluteOffset,
+                    total_buffer_length: &mut ev.TotalBufferLength,
+                    buffers: unsafe {
+                        std::slice::from_raw_parts(ev.Buffers, ev.BufferCount as usize)
```

Comment:
How many dynamic memory allocations does all this require? I assume Rust will dynamically allocate the `Self::Receive` and all the `slice`s here? Or is it able to use stack memory somehow? Bottom line, I see this as introducing a possibly significant bottleneck to a high throughput receive path. In MsQuic we do a LOT of work to completely eliminate dynamic allocations in the datapath, and I'd like to ensure that (somehow) maintains into Rust as well.

Reply:
This does not require dynamic alloc. "slice" in rust is like string_view in cpp (on stack). The goal or the requirement I have for the rust wrppers is zero dynamic alloc and zero copy for the buffers.

Reply:
How does it do it statically, when the number of slices is dynamic?

Reply:
buffer type is of this:
```rs
buffers: &'a [crate::ffi::QUIC_BUFFER],
```
The memory layout is just
```
struct{
  data: *const crate::ffi::QUIC_BUFFER
  len: usize
}
```
The advantage of using the slice is that u can use iterators etc utilites to traverse it, and makes the lifetime known to the compiler.
I have another branch where QUIC_BUFFER is wrapped and implemented iterators and lifetime etc as well.
The goal is to have wrapper types to expose &[u8] types to users, and the memory the slice points to are owned by msquic. (We will need to figure out for advanced use case where msquic extends the lifetime of the buffer)